### PR TITLE
Millenium patron authentication integration should support alternative patron identifiers

### PR DESCRIPTION
**Pull request by io7m**
## Description

This adds an option to the admin UI to specify a field that will be
used as a barcode instead of the default 'pb' field. Some systems
evidently use fields such as the username field ('pu') as the patron
identifier. This changes the code such that one other field can be
treated as a patron identifier, and the effect is additive. Specifically:

  * If a 'pb' field exists, it is added as a potential identifier.
  * If the configuration specifies that another field `x` should
    be treated as a patron identifier, it is also added as a potential
    identifier.
  * The field `x` will _also_ be processed exactly as it would have
    been before this change, so for example if the configuration
    specifies that the 'pu' field will be used as a patron identifier,
    it will _also_ be processed and used as a patron username.

## Motivation and Context

Some systems are evidently configured not to deliver `pb` "barcode"
fields. This allows the CM to work with those systems.

https://www.notion.so/lyrasis/Millenium-patron-authentication-integration-should-support-alternative-patron-identifiers-f1213e709eae4294aeccdb82a6abea94

## How Has This Been Tested?

New test cases have been added to show that both the configured field is
used as a patron identifier, and that the existing interpretations of fields aren't
disrupted.

## Checklist

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by tdilauro** [comment]
If `pu` is specified as the identifier field, that should allow `pu` to continue to be processed as a username.

File: api/millenium_patron.py [hunk 1]
```diff
@@ -385,18 +408,27 @@ def patron_dump_to_patrondata(self, current_identifier, content):
         potential_identifiers = []
         for f, v in self._extract_text_nodes(content):
             k = self._code_from_field(f)
-            if k == self.BARCODE_FIELD:
+
+            # Check to see if the key should be treated as if it is a patron identifier.
+            # This is dependent on a configuration setting. The key will also have a
+            # chance to be treated as a different kind of field below (for example, if
+            # the configuration says that the username field 'pu' should be treated as
+            # a patron identifier, we _also_ want to treat is as a username below; both
```
Comment (tdilauro) [inline api/millenium_patron.py:416]: minor: typo "...treat [is] as a username..." -> "...treat [it] as a username..."

File: tests/api/test_millenium_patron.py [hunk 2]
```diff
@@ -142,6 +148,24 @@ def test__remote_patron_lookup_success_nonsensical_labels(self):
...
 
+    def test__remote_patron_lookup_success_alternative_identifier(self):
+        self.api = self.mock_api(field_used_as_patron_identifier="pu")
+        self.api.enqueue("dump.success_alternative_identifier.html")
+        patrondata = PatronData(authorization_identifier="good barcode")
+        patrondata = self.api._remote_patron_lookup(patrondata)
+
+        # The barcode is correctly captured from the "NONSENSE[pb]" element.
+        # This checks that we care about the 'pb' code and don't care about
+        # what comes before it.
```
Comment (tdilauro) [inline tests/api/test_millenium_patron.py:159]: It looks like this comment was copied from the previous test and hasn't been updated to reflect the enqueued file or alternative `pu` identifier field for this test.

**Review by tdilauro — COMMENTED** [review]
This looks good. Just a couple of minor issues to address.

File: tests/api/test_millenium_patron.py [hunk 3]
```diff
@@ -142,6 +148,22 @@ def test__remote_patron_lookup_success_nonsensical_labels(self):
         assert "alice@sheldon.com" == patrondata.email_address
         assert PatronData.NO_VALUE == patrondata.block_reason
 
+    def test__remote_patron_lookup_success_alternative_identifier(self):
+        self.api = self.mock_api(field_used_as_patron_identifier="pu")
+        self.api.enqueue("dump.success_alternative_identifier.html")
+        patrondata = PatronData(authorization_identifier="good barcode")
+        patrondata = self.api._remote_patron_lookup(patrondata)
+
+        # The barcode is correctly captured from the "MENINX[pu]" element.
```
[the remark this answers is not in the document]
Comment (io7m) [reply tests/api/test_millenium_patron.py:157]: That's a mistake in the comment. By "barcode", I mean "identifier".

**Review by tdilauro — APPROVED** [review]
This looks good to go! Thanks!

**Comment by io7m** [comment]
I'll just wait for the tests to finish. :wink:

**Pull request merged**